Updates to latest Kafka, using a slimmer image and new healthcheck

adriancole opened the pull request:
This updates the Kafka+ZooKeeper image to latest 2.6 which allows JRE
14. It also prunes the extra content needed for services we don't need
like Kafka Connect and Streams. Finally, it updates the health-check per
recent testing in docker-compose.

This saves ~24M off the image total even after upgrading. Note: the total is the amount above the java image.
```
hypertrace/kafka-zookeeper                test                    9f564956290a        5 minutes ago       133MB
hypertrace/kafka-zookeeper                latest                  17c7d51b892a        23 hours ago        157MB
hypertrace/java                           11                      86afb574b9eb        11 days ago         101MB
```

adriancole commented on kafka-zookeeper/docker-healthcheck:
```diff
@@ -0,0 +1,12 @@
+#!/bin/sh
+set -eo pipefail
+
+host="$(hostname --ip-address || echo '127.0.0.1')"
```
filename and script convention per https://github.com/docker-library/healthcheck/blob/master/cassandra/docker-healthcheck

adriancole commented:
wait a bit as I accidentally didn't test the healthcheck as docker-compose was overriding it

adriancole commented:
ok should be good now

jeqo commented:
> @jeqo in case you know.. as far as I can tell, rocksdb is not invoked on the broker, it is how local state stores are managed even if there is a changelog topic. Hence, there's no broker need for rocksdb even if kstreams apps are in use, correct?
> 
> https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Streams+Internal+Data+Management

correct. broker is independent of streams and connect libs.

adriancole commented:
Thanks for the help @jeqo!

@buchi-busireddy since I've addressed your comment, can you change the status from changes requested to something not that :)

adriancole commented:
@ravisingal I will do a little cleanup then ping you for a review to clear this

adriancole commented:
I want to clean up the workarounds for the ancient Docker 17, so will wait for https://github.com/hypertrace/kafka/pull/9 to continue this, and also wait on that merging before fanning out the same change to NN repositories.

adriancole commented:
waiting on this https://github.com/hypertrace/helm-gcs-packager/pull/5

adriancole commented:
refactorings done. just waiting to figure out how to release helm packager, so can add to the updates here https://github.com/hypertrace/hypertrace/issues/19#issuecomment-679814145

ravisingal commented on kafka-zookeeper/Dockerfile:
```diff
@@ -1,42 +1,39 @@
 MAINTAINER Hypertrace "https://www.hypertrace.org/"
 
+# Add HEALTHCHECK and ENTRYPOINT scripts into the default search path
+COPY docker/collector/kafka/docker-bin/* /usr/local/bin/
```
check source path. it should be `docker-bin/*`.

adriancole replied:
indeed fixed and tested it locally.

adriancole commented:
re-verified image size

```
hypertrace/kafka-zookeeper                test                                bc7bf342e395        29 seconds ago      133MB
hypertrace/kafka-zookeeper                latest                              17c7d51b892a        5 days ago          157MB
```

aaron-steinfeld commented on build.gradle.kts:
```diff
@@ -1,6 +1,6 @@
 plugins {
-  id("org.hypertrace.docker-plugin") version "0.5.1"
-  id("org.hypertrace.docker-publish-plugin") version "0.5.1"
+  id("org.hypertrace.docker-plugin") version "0.6.0"
```
This will want to move to 0.6.1 for the same empty branch issue we saw in https://app.circleci.com/pipelines/github/hypertrace/helm-gcs-packager/29/workflows/aba305d1-6051-405a-a397-85a7b5b91b81/jobs/33

adriancole replied:
ah good eye!

aaron-steinfeld replied:
I guess it technically doesn't matter unless we're making the circleci changes here too, since that's only encountered on a tag-triggered release.

adriancole commented:
verified in docker compose

```
± |updates ✓| → docker ps |grep kafka-zookeeper
15b4a980dc68        hypertrace/kafka-zookeeper:test         "start-kafka-zookeep…"   28 seconds ago      Up 27 seconds (healthy)            2181/tcp, 9092/tcp, 19092/tcp                                             kafka-zookeeper

```

adriancole commented:
I want to fix the health output .. one sec.

```
$ docker inspect --format='{{json .State.Health}}' kafka-zookeeper |jq .
{
  "Status": "healthy",
  "FailingStreak": 0,
  "Log": [
    {
      "Start": "2020-08-26T01:25:34.8925409Z",
      "End": "2020-08-26T01:25:34.9781041Z",
      "ExitCode": 0,
      "Output": "ruok is not executed because it is not in the whitelist.\n"
    },
    {
      "Start": "2020-08-26T01:25:35.9814607Z",
      "End": "2020-08-26T01:25:36.0638757Z",
      "ExitCode": 0,
      "Output": "ruok is not executed because it is not in the whitelist.\n"
    },
    {
      "Start": "2020-08-26T01:25:37.0706208Z",
      "End": "2020-08-26T01:25:37.1612689Z",
      "ExitCode": 0,
      "Output": "ruok is not executed because it is not in the whitelist.\n"
    },
    {
      "Start": "2020-08-26T01:25:38.1669986Z",
      "End": "2020-08-26T01:25:38.2464031Z",
      "ExitCode": 0,
      "Output": "ruok is not executed because it is not in the whitelist.\n"
    },
    {
      "Start": "2020-08-26T01:25:39.2529175Z",
      "End": "2020-08-26T01:25:39.3345667Z",
      "ExitCode": 0,
      "Output": "ruok is not executed because it is not in the whitelist.\n"
    }
  ]
}
```

adriancole commented:
ok health is overhauled and prints decent status messages

On fresh start
```
docker inspect --format='{{json .State.Health}}' kafka-zookeeper|jq .
{
  "Status": "healthy",
  "FailingStreak": 0,
  "Log": [
    {
      "Start": "2020-08-26T03:42:23.0325193Z",
      "End": "2020-08-26T03:42:23.1128395Z",
      "ExitCode": 1,
      "Output": "Kafka port check failed with code: 1\n"
    },
    {
      "Start": "2020-08-26T03:42:24.1168101Z",
      "End": "2020-08-26T03:42:24.212042Z",
      "ExitCode": 0,
      "Output": ""
    },
    {
      "Start": "2020-08-26T03:42:25.1865672Z",
      "End": "2020-08-26T03:42:25.2850543Z",
      "ExitCode": 0,
      "Output": ""
    },
    {
      "Start": "2020-08-26T03:42:26.2903133Z",
      "End": "2020-08-26T03:42:26.3809064Z",
      "ExitCode": 0,
      "Output": ""
    },
    {
      "Start": "2020-08-26T03:42:27.3899288Z",
      "End": "2020-08-26T03:42:27.4825764Z",
      "ExitCode": 0,
      "Output": ""
    }
  ]
}
```


After killing pid 1
```
 docker inspect --format='{{json .State.Health}}' kafka-zookeeper|jq .
{
  "Status": "healthy",
  "FailingStreak": 2,
  "Log": [
    {
      "Start": "2020-08-26T03:44:46.2733427Z",
      "End": "2020-08-26T03:44:46.3523355Z",
      "ExitCode": 0,
      "Output": ""
    },
    {
      "Start": "2020-08-26T03:44:47.3604469Z",
      "End": "2020-08-26T03:44:47.4432535Z",
      "ExitCode": 0,
      "Output": ""
    },
    {
      "Start": "2020-08-26T03:44:48.4481181Z",
      "End": "2020-08-26T03:44:48.5284113Z",
      "ExitCode": 0,
      "Output": ""
    },
    {
      "Start": "2020-08-26T03:44:49.5336821Z",
      "End": "2020-08-26T03:44:49.6250508Z",
      "ExitCode": 1,
      "Output": "Kafka port check failed with code: 1\n"
    },
    {
      "Start": "2020-08-26T03:44:50.6299007Z",
      "End": "2020-08-26T03:44:50.7163721Z",
      "ExitCode": 1,
      "Output": "Kafka port check failed with code: 1\n"
    }
  ]
}
```

After killing zookeeper
```
docker inspect --format='{{json .State.Health}}' kafka-zookeeper|jq .
{
  "Status": "unhealthy",
  "FailingStreak": 3,
  "Log": [
    {
      "Start": "2020-08-26T03:44:03.4278552Z",
      "End": "2020-08-26T03:44:03.5127206Z",
      "ExitCode": 0,
      "Output": ""
    },
    {
      "Start": "2020-08-26T03:44:04.5178278Z",
      "End": "2020-08-26T03:44:04.597353Z",
      "ExitCode": 0,
      "Output": ""
    },
    {
      "Start": "2020-08-26T03:44:05.6051764Z",
      "End": "2020-08-26T03:44:05.6925962Z",
      "ExitCode": 1,
      "Output": "ZooKeeper port check failed with code: 1\n"
    },
    {
      "Start": "2020-08-26T03:44:06.6972305Z",
      "End": "2020-08-26T03:44:06.7810542Z",
      "ExitCode": 1,
      "Output": "ZooKeeper port check failed with code: 1\n"
    },
    {
      "Start": "2020-08-26T03:44:07.7868709Z",
      "End": "2020-08-26T03:44:07.8657205Z",
      "ExitCode": 1,
      "Output": "ZooKeeper port check failed with code: 1\n"
    }
  ]
}
```

adriancole commented on kafka-zookeeper/Dockerfile:
```diff
@@ -1,42 +1,40 @@
-  # We use bootstrap-server, not zookeeper, as KIP-500 will eventually remove the dependency
-  CMD /opt/kafka/bin/kafka-topics.sh --bootstrap-server 127.0.0.1:9092 --list > /dev/null || exit 1
+# We use start period of 30s to avoid marking Kafka unhealthy on slow or contended CI hosts
+HEALTHCHECK --interval=1s --start-period=30s --timeout=5s CMD ["docker-healthcheck"]
```
@hypertrace/build-tooling @jcchavezs @JBAhire note: I'm going to use `--start-period=30s --timeout=5s ` everywhere because we can't predict laptop freezes, and using 5s timeout allows this to not interfere

JBAhire replied:
makes sense!

adriancole commented on kafka-zookeeper/docker-bin/docker-healthcheck:
```diff
@@ -0,0 +1,31 @@
+# HEALTHCHECK for use in `docker ps` or `docker-compose ps`.
+# It can also be used as a readiness probe in k8s
+
+# Fail on unset variables, but don't quit on rc!=0, so we can log what happened
```
This is the script re-written and I'll do similarly everywhere something is a composite service or otherwise blamed.. ahem.. pinot.. ahem

ravisingal commented on kafka-zookeeper/install:
```diff
@@ -0,0 +1,77 @@
+# don't log ZK connections
+sed -i 's~log4j.logger.org.apache.zookeeper=INFO~log4j.logger.org.apache.zookeeper=WARN~g' config/log4j.properties
+# allow ruok command for testing ZK health
+echo 4lw.commands.whitelist=srvr, ruok >> config/zookeeper.properties
```
we can remove this? basic zookeeper configuration are set later in the script.

adriancole replied:
yep. this was a relic

adriancole commented:
rechecked with docker-compose last commit including health output

adriancole commented:
Thanks for the look folks!
